// @not-govuk/components/dist/components/button/index.js
import Start from './start';
import Anchor from './anchor';
import Input from './input';
import { className } from '../../helpers';
;
/**
 * Use `Button` to give people something to click on.
 */
export const Button = props => {
    const type = props.submit ? 'submit' : 'button';
    const anchor = props.href !== null;
    const classes = className('govuk-button', props.disabled ? 'govuk-button--disabled' : '', props.secondary ? 'govuk-button--secondary' : '', props.warning ? 'govuk-button--warning' : '', props.className);
    const text = props.value || (props.start && 'Start now') || undefined;
    const processedProps = {
        ...props,
        text: text,
        type: type,
        className: classes
    };
    return (props.start
        ? Start(processedProps)
        : (anchor
            ? Anchor(processedProps)
            : Input(processedProps)));
};
Button.defaultProps = {
    className: null,
    disabled: false,
    href: null,
    id: null,
    onClick: null,
    secondary: false,
    start: false,
    submit: false,
    value: null,
    warning: false
};
export const StartButton = props => Button({ ...Button.defaultProps, ...props, start: true });
export const SubmitButton = props => Button({ ...Button.defaultProps, ...props, submit: true });
export default Button;
