// @not-govuk/components/dist/components/date-input/index.js
import { createElement as h } from 'react';
import FormField from '../form-field';
;
export const DateInput = props => h(FormField, {
    autoComplete: props.autoComplete,
    className: props.className,
    defaultValue: props.defaultValue,
    disabled: props.disabled,
    error: props.error,
    hint: props.hint,
    id: props.id,
    label: props.label,
    name: props.name,
    onBlur: props.onBlur,
    onChange: props.onChange,
    type: 'date',
    value: props.value
});
DateInput.defaultProps = {
    autoComplete: null,
    className: null,
    defaultValue: undefined,
    disabled: false,
    error: null,
    hint: null,
    id: null,
    onBlur: null,
    onChange: null,
    value: undefined
};
DateInput.format = (v) => {
    const pad = (size, v) => String(v).padStart(size, '0');
    const isSet = (v) => !!(v || v === 0);
    if (isSet(v.day) && isSet(v.month) && isSet(v.year)) {
        const dd = pad(2, v.day);
        const mm = pad(2, v.month);
        const yyyy = pad(4, v.year);
        return `${yyyy}-${mm}-${dd}`;
    }
    else {
        return undefined;
    }
};
DateInput.deformat = (v) => {
    const unpad = (v) => Number(v).toString();
    const arr = v.split('-');
    return (arr.length === 3 ? {
        day: unpad(arr[2]),
        month: unpad(arr[1]),
        year: arr[0]
    } : undefined);
};
export default DateInput;
