// @not-govuk/forms/dist/graph/field.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isFieldItem = exports.isFieldNode = exports.FieldNode = exports.FieldItem = void 0;
const node_1 = require("./node");
class FieldItem {
    constructor(active, name) {
        this.tag = 'field';
        this.active = active;
        this.name = name;
    }
}
exports.FieldItem = FieldItem;
class FieldNode extends node_1.Node {
    constructor(name, format, validate, preValidate) {
        super('field');
        this.name = name;
        this.format = format;
        this.preValidate = preValidate;
        this.validate = validate;
        this.hasFocus = false;
    }
    toPathItem() {
        return new FieldItem(this.hasFocus, this.name);
    }
    ;
    depopulate() {
        this.hasFocus = false;
    }
    populateFromValues(values) {
    }
    populateFromNext(next) {
        this.hasFocus = this.name === next;
    }
    toArray() {
        return [this];
    }
    traverse(values) {
        return [this];
    }
}
exports.FieldNode = FieldNode;
exports.isFieldNode = (v) => v instanceof FieldNode;
exports.isFieldItem = (v) => v instanceof FieldItem;
