// @not-govuk/forms/dist/graph/page.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPageItem = exports.isPageNode = exports.PageNode = exports.PageItem = void 0;
const graph_1 = require("./graph");
const node_1 = require("./node");
class PageItem {
    constructor(active) {
        this.tag = 'page';
        this.active = active;
    }
}
exports.PageItem = PageItem;
class PageNode extends node_1.Node {
    constructor(contents) {
        super('page');
        this.contents = contents;
        this.active = true;
    }
    toPathItem() {
        return new PageItem(this.active);
    }
    ;
    depopulate() {
        this.fields = [];
        this.active = true;
    }
    populateFromValues(values) {
        this.fields = this.contents.gatherFieldsAlongPath(values);
    }
    populateFromNext(next) {
        this.active = this.fields.includes(next);
    }
    toArray() {
        return [
            this,
            ...this.contents.toArray()
        ];
    }
    traverse(values) {
        return (this.active
            ? new graph_1.Graph(this, ...this.contents.flatMap(e => e.traverse(values)))
            : new graph_1.Graph(this));
    }
}
exports.PageNode = PageNode;
exports.isPageNode = (v) => v instanceof PageNode;
exports.isPageItem = (v) => v instanceof PageItem;
