// @not-govuk/engine/dist/lib/pages.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.pageRoutes = exports.gatherPages = void 0;
const restify_1 = require("@not-govuk/restify");
const path_1 = __importDefault(require("path"));
const pageExtensionPattern = /\.([jt]sx?|html)$/i;
const removePrecedingDotSlash = (s) => (s.startsWith('./')
    ? s.slice(2)
    : s);
const removeTrailingSlash = (s) => (s.endsWith('/')
    ? s.slice(0, -1)
    : s);
const addPreceedingSlash = (s) => (s.startsWith('/')
    ? s
    : '/' + s);
const src2Href = (page) => (addPreceedingSlash(removePrecedingDotSlash(page)
    .replace(pageExtensionPattern, '')
    .replace(/index$/, '')
    .split(path_1.default.sep)
    .map(encodeURI)
    .join('/')));
const capitaliseFirstLetter = ([x, ...xs]) => ([x.toUpperCase(), ...xs].join(''));
const src2Title = (page) => (capitaliseFirstLetter(page
    .split('/')
    .pop()
    .split('.')
    .shift()));
const href2Path = (s) => (addPreceedingSlash(removeTrailingSlash(s)));
exports.gatherPages = (pageLoader) => Promise.all(pageLoader
    .keys()
    .map(async (e) => {
    const mod = await pageLoader(e);
    return {
        Component: (typeof mod === 'string'
            ? mod
            : mod.default),
        href: src2Href(e),
        src: e,
        title: (typeof mod === 'string'
            ? src2Title(e)
            : mod.title)
    };
}));
const pageMiddleware = (title) => (req, res, next) => {
    res.renderApp(200, title).finally(next);
};
exports.pageRoutes = (pages) => {
    const router = new restify_1.Router();
    pages.forEach(e => router.get(href2Path(e.href), pageMiddleware(e.title)));
    return router;
};
