// @not-govuk/engine/dist/lib/webpack.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.webpackMiddleware = void 0;
const webpack_1 = __importDefault(require("webpack"));
const webpack_dev_middleware_1 = __importDefault(require("webpack-dev-middleware"));
const webpack_hot_middleware_1 = __importDefault(require("webpack-hot-middleware"));
const etag_1 = __importDefault(require("etag"));
exports.webpackMiddleware = webpackConfig => {
    const compiler = webpack_1.default(webpackConfig);
    const devOptions = {
        stats: webpackConfig.stats,
        publicPath: webpackConfig.output.publicPath,
        serverSideRender: true
    };
    const dev = webpack_dev_middleware_1.default(compiler, devOptions);
    const hot = webpack_hot_middleware_1.default(compiler, {});
    return {
        serveFiles: (req, res, next) => {
            res.locals = res.locals || {};
            const restifyTransport = {
                setHeader(key, val) {
                    res.setHeader(key, val);
                },
                send(content) {
                    res.charSet('utf-8');
                    res.writeHead(this.statusCode || 200, {
                        'Cache-Control': 'private, no-cache',
                        'ETag': etag_1.default(content, { weak: true })
                    });
                    res.write(content);
                    res.end();
                },
                end() {
                    return;
                },
                locals: res.locals
            };
            return dev(req, restifyTransport, next);
        },
        hot,
        hotPath: '/__webpack_hmr'
    };
};
exports.default = exports.webpackMiddleware;
