// @not-govuk/components/dist/forms/index.js
import { createElement as h } from 'react';
import WTForm, { Fork, Page, alpha as localAlpha, alphanumeric as localAlphanumeric, mobileNumber as localMobileNumber, postalCode as localPostalCode } from '@not-govuk/forms';
import { Submit } from './controls';
import { Checkboxes, DateInput, Field, Radios, Select, TextInput, Textarea } from './fields';
export const FormComponent = ({ classBlock, ...props }) => h(WTForm, {
    ...props,
    classBlock: classBlock || 'not-govuk-form'
});
export const Form = Object.assign(FormComponent, {
    Checkboxes,
    DateInput,
    Field,
    Fork,
    Page,
    Radios,
    Select,
    Submit,
    TextInput,
    Textarea
});
const defaultLanguage = 'en';
const defaultCountry = 'GB';
const defaultLocale = `${defaultLanguage}_${defaultCountry}`;
export const alpha = localAlpha(defaultLocale);
export const alphanumeric = localAlphanumeric(defaultLocale);
export const mobileNumber = localMobileNumber(defaultLocale);
export const postcode = localPostalCode(defaultCountry);
export default Form;
export { localAlpha, localAlphanumeric, localMobileNumber, localPostalCode };
export { after, before, date, email, exactLength, future, integer, maximum, maxLength, maxWords, minimum, minLength, minWords, numeric, past, range, required, url, validator, withControl, withForm, withField } from '@not-govuk/forms';
